consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java

```diff
@@ -0,0 +1,137 @@
+ */
+public class RoundChangeManager {
+
+  public static class RoundChangeStatus {
+    private final int quorumSize;
+
+    @VisibleForTesting
+    final Map<Address, SignedData<RoundChangePayload>> receivedMessages = Maps.newHashMap();
+
+    private boolean actioned = false;
+
+    public RoundChangeStatus(final int quorumSize) {
+      this.quorumSize = quorumSize;
+    }
+
+    public void addMessage(final SignedData<RoundChangePayload> msg) {
+      if (!actioned) {
+        receivedMessages.put(msg.getSender(), msg);
+      }
+    }
+
+    public boolean roundChangeReady() {
+      return receivedMessages.size() >= quorumSize && !actioned;
+    }
+
+    public RoundChangeCertificate createRoundChangeCertificate() {
+      if (roundChangeReady()) {
+        actioned = true;
+        return new RoundChangeCertificate(receivedMessages.values());
+      } else {
+        throw new IllegalStateException("Unable to create RoundChangeCertificate at this time.");
+      }
+    }
+  }
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  @VisibleForTesting
+  final Map<ConsensusRoundIdentifier, RoundChangeStatus> roundChangeCache = Maps.newHashMap();
+
+  private final Collection<Address> validators;
+  private final int quorumSize;
+  private final long sequenceNumber;
+  private final MessageValidatorFactory messageValidityFactory;
+
+  public RoundChangeManager(
+      final long sequenceNumber,
+      final Collection<Address> validators,
+      final MessageValidatorFactory messageValidityFactory) {
+    this.validators = validators;
+    this.messageValidityFactory = messageValidityFactory;
+    this.sequenceNumber = sequenceNumber;
+    this.quorumSize = IbftHelpers.calculateRequiredValidatorQuorum(validators.size());
+  }
+
+  public Optional<RoundChangeCertificate> appendRoundChangeMessage(
+      final SignedData<RoundChangePayload> msg) {
+    // validate the received msg
+    final ConsensusRoundIdentifier msgTargetRound = msg.getPayload().getRoundChangeIdentifier();
+    final RoundChangeMessageValidator messageValidator =
+        new RoundChangeMessageValidator(
+            messageValidityFactory, validators, quorumSize, sequenceNumber);
+    if (!messageValidator.validateMessage(msg)) {
+      LOG.info("RoundChange message was invalid.");
+      return Optional.empty();
+    }
+    if (!roundChangeCache.containsKey(msgTargetRound)) {
+      roundChangeCache.put(msgTargetRound, new RoundChangeStatus(quorumSize));
+    }
+    final RoundChangeStatus roundChangeStatus = roundChangeCache.get(msgTargetRound);
+    roundChangeStatus.addMessage(msg);
+    if (roundChangeStatus.roundChangeReady()) {
+      return Optional.of(roundChangeStatus.createRoundChangeCertificate());
+    }
+    return Optional.empty();
+  }
+
+  /* Called when a given round has completed/timed-out such that msgs cached can be removed */
+  public void discardCompletedRound(final ConsensusRoundIdentifier completedRoundIdentifier) {
```

Comment:
Checkout using Collection.removeif - probably makes this a one-liner.
roundChangeCache.entrySet().removeIf(e -> e.getKey)

Reply:
done